### ISSUE TITLE
Cannot import cupy-cuda111 on Windows, CUDA 11.1

### ISSUE THREAD

**Issue by nvAli** [cupy/cupy#4224]
Hi, installation of cupy does fail on my system - Windows 10, CUDA 11.1, driver 460.40, Python 3.8.

I installed cupy via
> pip install cupy-cuda111

and my fresh Python 3.8 installation shows
> C:\Users\afritzsche>pip freeze
  cupy-cuda111==8.1.0
  fastrlock==0.5
  numpy==1.19.4

But trying to import cupy fails with
> DLL load failed while importing cusolver: The specified module could not be found.

The complete output is attached
[log.txt](https://github.com/cupy/cupy/files/5484301/log.txt)

I tried already the solution proposed here
  https://github.com/cupy/cupy/issues/2243#issuecomment-543419082
hence the output of the DLL path in the attached log.txt. But it didn't changed anything and the cupy import still fails.

Thanks for any help,
Ali

**Comment by emcastillo** [comment]
Have you set the CUDA_PATH env var?

**Comment by nvAli** [comment]
Yes. Hence the output in the log.txt from the additional code in __init.py__
`print('Add DLL path {}'.format(os.path.join(os.environ['CUDA_PATH'], 'bin')))`
> Add DLL path C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\bin
Could somehow the spaces confuse the handling? But then I probably wouldn't be the first reporting that...

**Comment by nvAli** [comment]
And looking at the source causing this failure in device.pyx and the line right before it
`from cupy_backends.cuda.libs import cublas
from cupy_backends.cuda.libs import cusolver`
one can see that while cublas was apparently imported properly it fails with cusolver. I am not familiar with the internals of cupy but maybe the issue is within the cusolver backend itself?

**Comment by nvAli** [comment]
As a WAR I found out that building cupy myself allowed me to import it properly. Ie
> pip install cupy

did work for me.

As a side note: building it took quite some time while my machine was mostly idling. It seems as if there is no parallel compilation happening - maybe that could be improved?

**Comment by anaruse** [comment]
> As a side note: building it took quite some time while my machine was mostly idling. It seems as if there is no parallel compilation happening - maybe that could be improved?

You can reduce the build time by limiting the target GPU architecture as follows.

    $ export CUPY_NVCC_GENERATE_CODE=arch=compute_70,code=sm_70

https://docs.cupy.dev/en/stable/contribution.html#tips-for-developers

**Comment by leofang** [comment]
> > As a side note: building it took quite some time while my machine was mostly idling. It seems as if there is no parallel compilation happening - maybe that could be improved?
> 
> You can reduce the build time by limiting the target GPU architecture as follows.
> 
> ```
> $ export CUPY_NVCC_GENERATE_CODE=arch=compute_70,code=sm_70
> ```
> 
> https://docs.cupy.dev/en/stable/contribution.html#tips-for-developers

Let me expand a little on @anaruse's reply, and why we haven't done parallel builds so far. 

When compiling CuPy's modules, the two most time-consuming ones are `cupy.cuda.thrust` and `cupy.cuda.cub`, because they have many C++ template kernels and we request many specializations. Compared to them, compiling other modules is really nothing. So even if we parallelize the build, in the end we still pay significant time to these two. 

The reason that `CUPY_NVCC_GENERATE_CODE` could help a lot is because by default we also build for all SMs that we support, so the time is increased roughly linearly in the number of SMs. Therefore, if you only build for your own usage, you should only target your device. Apparently, no modules contain device code that is relevant to `CUPY_NVCC_GENERATE_CODE` other than the two notorious modules 😁

**Comment by nvAli** [comment]
Thanks for your explanations. I don't mind some compile times (and am aware of the 'for all SMs' vs 'just for mine' build issue) and added it hence just as a side note. But I wasn't aware that Thrust and CUB are the bottlenecks here.

But I guess that the issue with the CUDA 11.1 Windows pre-build cupy package is more serious and probably others will hit it soon. Assuming that I am not the sole cupy Windows user ;-)

**Comment by kmaehashi** [comment]
So far I'm unable to reproduce this issue. Could you confirm that `cusolver64_11.dll` exists under `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\bin`?

FYI, here are DLL dependents:

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community>dumpbin /dependents C:\Development\Python\Python38\Lib\site-packages\cupy_backends\cuda\libs\cusolver.cp38-win_amd64.pyd
Microsoft (R) COFF/PE Dumper Version 14.16.27027.1
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file C:\Development\Python\Python38\Lib\site-packages\cupy_backends\cuda\libs\cusolver.cp38-win_amd64.pyd

File Type: DLL

  Image has the following dependencies:

    cusolver64_11.dll
    python38.dll
    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-runtime-l1-1-0.dll

  Summary

        C000 .data
        1000 .gfids
        3000 .pdata
       1F000 .rdata
        2000 .reloc
       6B000 .text

```

**Comment by nvAli** [comment]
Good catch. In my installation it indeed does not exist, but only a cusolver64_10.dll. Wondering why you aren't seeing that issue I looked into the most up-to-date CUDA 11.1 Update 1 installer and found that while there is now a cusolver64_11.dll there are still a few dlls like cufft64_10.dll and curand64_10.dll.

So, definitely not a cupy issue but a CUDA installer one. But maybe you could add it to some Windows Installation notes/readme? I will close the issue now, thanks.

**Comment by nvAli** [comment]
It actually seems as if this is an intended change in the installer. I will follow that issue and report back. And might, dependent of the outcome, reopen the ticket again.

**Comment by nvAli** [comment]
Re-opening. As of CUDA 11 the versioning of its components is actually a feature
  https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions
So, the current approach of cupy is not reliably working anymore for all CUDA >= 11.0.

**Comment by kmaehashi** [comment]
Thank you for the information!

* CUDA 11.1.0 contains cuSOLVER 11.0.0.74
* CUDA 11.1.1 (Update 1) contains cuSOLVER 11.0.1.105

So I guess it was a bug in CUDA 11.1.0 that mistakenly shipped cuSOLVER 11 with the name `cusolver64_10.dll`.

That said,

> As of CUDA 11 the versioning of its components is actually a feature
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions

I think many Python projects shipping wheels like CuPy assume shared library names do not change between minor CUDA releases (11.1 and 11.1 Update 1). Should we change this assumption? cc/ @anaruse

**Comment by nvAli** [comment]
Maybe there wasn't a bug int the CUDA 11.1.0. I had spent some time but couldn't find out where I had my installed CUDA 11.1.0 installer from - potentially it was an internal one. Looking into both publicly available 11.1.0 and 11.1.1 I see what you already stated - the cuSolver dll hasn't changed the filename.

So, maybe there is no need to guard against filename changes between CUDA minor release. While I haven't found any promise that CUDA minor releases won't change component file names I have no real evidence neither (besides my local installation - which shouldn't count as long as there is no other complaing). So, feel free to ignore this issue - I can close it if you'd like.

**Comment by nvAli** [comment]
I will close the issue for now. It is not clear whether it was caused by a public CUDA toolkit.

So, in case that there are similar issue reports are coming in you might already know what could be the culprit. Until then one probably can rely on identical component names between CUDA minor releases.
